Listing locals in block of var declarations panics

This issue is potentially related to #83 

While stepping through the `http.Client.Do()` call, I tried listing the local variables on line 498 (golang 1.8.3 docker image). I had just installed delve 1.0.0-rc.2 (full output of `{dlv,go} version` below). This resulted in a runtime panic (index out of range) probably when trying to pretty-print variables that hadn't been initialised at that point.

Going by the output it seems that the problem is that I was exactly on the line that declares `reqs        []*Request`. Variables later on that haven't been initialised do show up with an error message (for example `copyHeaders = (unreadable input/output error)`)

I'm not sure what the best course of action is, but I suspect there needs to be a nil check on slices like this, and display its value as `nil`, or even with the same error message as `copyHeaders` had. 

The relevant information below:

## Versions:

Delve:

```
dlv version
Delve Debugger
Version: 1.0.0-rc.2
Build: $Id: 5c7676404d10d52e16be57fea744a7312b831f0c $
```

Go

```
go version
go version go1.8.3 linux/amd64
```

OS/arch

docker image golang:1.8.3 (Debian jessie based):

```
# cat /etc/*-release
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

# uname -a
Linux 6adc032f03e8 3.10.0-514.26.2.el7.x86_64 #1 SMP Tue Jul 4 15:04:05 UTC 2017 x86_64 GNU/Linux
```

## Delve trace

```
(dlv) s
> net/http.(*Client).Do() /usr/local/go/src/net/http/client.go:498 (PC: 0x5a3c18)
   493:			return nil, errors.New("http: nil Request.URL")
   494:		}
   495:	
   496:		var (
   497:			deadline    = c.deadline()
=> 498:			reqs        []*Request
   499:			resp        *Response
   500:			copyHeaders = c.makeHeadersCopier(req)
   501:	
   502:			// Redirect behavior:
   503:			redirectMethod string
(dlv) locals
includeBody = false
t·2.nsec = 0
t·2.sec = 0
redirectMethod.len = 842350530224
loc.len = 0
ue.itab = (*uint8)(0x8)
ue.data = (*uint8)(0x7fe083b2a000)
u = *net/url.URL nil
t·2.loc = *time.Location nil
redirectMethod.ptr = (*uint8)(0xc420043ae0)
loc.ptr = (*uint8)(0x40ff12)
ireq = (*net/http.Request)(0xc420060c60)
err.itab = (*uint8)(0x30)
err.itab#13 = (*uint8)(0x30)
err.data = (*uint8)(0x640040)
err.data#15 = (*uint8)(0x610bc0)
didTimeout = (unreadable input/output error)
copyHeaders = (unreadable input/output error)
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/derekparker/delve/service/api.(*Variable).writeTo(0xc4207e35c0, 0xc8ce00, 0xc4205558f0, 0xc420010001, 0x0, 0x0)
	/go/src/github.com/derekparker/delve/service/api/prettyprint.go:54 +0x10e9
github.com/derekparker/delve/service/api.(*Variable).SinglelineString(0xc4207e35c0, 0xc8cc80, 0x1)
	/go/src/github.com/derekparker/delve/service/api/prettyprint.go:20 +0x78
github.com/derekparker/delve/pkg/terminal.printFilteredVariables(0xa10e56, 0x6, 0xc420063500, 0x15, 0x1c, 0x0, 0x0, 0x0, 0x0, 0x40, ...)
	/go/src/github.com/derekparker/delve/pkg/terminal/command.go:1021 +0x1bc
github.com/derekparker/delve/pkg/terminal.locals(0xc4204f2960, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0x0, 0x0, 0x2, 0x6)
	/go/src/github.com/derekparker/delve/pkg/terminal/command.go:1096 +0x2c9
github.com/derekparker/delve/pkg/terminal.(*Commands).CallWithContext(0xc420572ed0, 0xc4204d5364, 0x6, 0xc4204f2960, 0x0, 0xffffffffffffffff, 0x0, 0x0, 0x6, 0xc4204d5364)
	/go/src/github.com/derekparker/delve/pkg/terminal/command.go:339 +0x134
github.com/derekparker/delve/pkg/terminal.(*Commands).Call(0xc420572ed0, 0xc4204d5364, 0x6, 0xc4204f2960, 0x0, 0x0)
	/go/src/github.com/derekparker/delve/pkg/terminal/command.go:344 +0x71
github.com/derekparker/delve/pkg/terminal.(*Term).Run(0xc4204f2960, 0x0, 0x0, 0x0)
	/go/src/github.com/derekparker/delve/pkg/terminal/terminal.go:132 +0x38f
github.com/derekparker/delve/cmd/dlv/cmds.execute(0x0, 0xc42011c8b0, 0x1, 0x1, 0xc42011e2c0, 0x0, 0x0, 0x1, 0x0)
	/go/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:537 +0x94f
github.com/derekparker/delve/cmd/dlv/cmds.debugCmd.func1(0xc420142900, 0xc42011c670, 0x1, 0x1, 0x0)
	/go/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:300 +0x41d
github.com/derekparker/delve/cmd/dlv/cmds.debugCmd(0xc420142900, 0xc42011c670, 0x1, 0x1)
	/go/src/github.com/derekparker/delve/cmd/dlv/cmds/commands.go:301 +0x49
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).execute(0xc420142900, 0xc42011c610, 0x1, 0x1, 0xc420142900, 0xc42011c610)
	/go/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:647 +0x234
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420142240, 0xc420142b40, 0xc420142900, 0xc4201426c0)
	/go/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:733 +0x339
github.com/derekparker/delve/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420142240, 0xc420142240, 0xa0eedf)
	/go/src/github.com/derekparker/delve/vendor/github.com/spf13/cobra/command.go:692 +0x2b
main.main()
	/go/src/github.com/derekparker/delve/cmd/dlv/main.go:18 +0x91
```

--- a/_fixtures/clientdo.go
+++ b/_fixtures/clientdo.go
@@ -1,0 +1,10 @@
+package main
+
+import (
+	"net/http"
+	"net/url"
+)
+
+func main() {
+	http.DefaultClient.Do(&http.Request{URL: &url.URL{}})
+}

--- a/pkg/proc/variables.go
+++ b/pkg/proc/variables.go
@@ -1938,6 +1938,9 @@ func (scope *EvalScope) variablesByTag(tag dwarf.Tag, cfg *LoadConfig) ([]*Varia
 	for i, v := range vars {
 		if name := v.Name; len(name) > 1 && name[0] == '&' {
 			v = v.maybeDereference()
+			if v.Addr == 0 {
+				v.Unreadable = fmt.Errorf("no address for escaped variable")
+			}
 			v.Name = name[1:]
 			v.Flags |= VariableEscaped
 			vars[i] = v

--- a/service/api/prettyprint.go
+++ b/service/api/prettyprint.go
@@ -49,7 +49,7 @@ func (v *Variable) writeTo(buf io.Writer, top, newlines, includeType bool, inden
 	case reflect.Array:
 		v.writeArrayTo(buf, newlines, includeType, indent)
 	case reflect.Ptr:
-		if v.Type == "" {
+		if v.Type == "" || len(v.Children) == 0 {
 			fmt.Fprint(buf, "nil")
 		} else if v.Children[0].OnlyAddr && v.Children[0].Addr != 0 {
 			fmt.Fprintf(buf, "(%s)(0x%x)", v.Type, v.Children[0].Addr)
@@ -58,7 +58,11 @@ func (v *Variable) writeTo(buf io.Writer, top, newlines, includeType bool, inden
 			v.Children[0].writeTo(buf, false, newlines, includeType, indent)
 		}
 	case reflect.UnsafePointer:
-		fmt.Fprintf(buf, "unsafe.Pointer(0x%x)", v.Children[0].Addr)
+		if len(v.Children) == 0 {
+			fmt.Fprintf(buf, "unsafe.Pointer(nil)")
+		} else {
+			fmt.Fprintf(buf, "unsafe.Pointer(0x%x)", v.Children[0].Addr)
+		}
 	case reflect.String:
 		v.writeStringTo(buf)
 	case reflect.Chan:

--- a/service/test/variables_test.go
+++ b/service/test/variables_test.go
@@ -976,3 +976,28 @@ func TestConstants(t *testing.T) {
 		}
 	})
 }
+
+func setFunctionBreakpoint(p proc.Process, fname string) (*proc.Breakpoint, error) {
+	addr, err := proc.FindFunctionLocation(p, fname, true, 0)
+	if err != nil {
+		return nil, err
+	}
+	return p.SetBreakpoint(addr, proc.UserBreakpoint, nil)
+}
+
+func TestIssue1075(t *testing.T) {
+	withTestProcess("clientdo", t, func(p proc.Process, fixture protest.Fixture) {
+		_, err := setFunctionBreakpoint(p, "net/http.(*Client).Do")
+		assertNoError(err, t, "setFunctionBreakpoint")
+		assertNoError(proc.Continue(p), t, "Continue()")
+		for i := 0; i < 10; i++ {
+			scope, err := proc.GoroutineScope(p.CurrentThread())
+			assertNoError(err, t, fmt.Sprintf("GoroutineScope (%d)", i))
+			vars, err := scope.LocalVariables(pnormalLoadConfig)
+			assertNoError(err, t, fmt.Sprintf("LocalVariables (%d)", i))
+			for _, v := range vars {
+				api.ConvertVar(v).SinglelineString()
+			}
+		}
+	})
+}

